apps/worker/src/monitor/tx.status.monitor.service.ts

```diff
@@ -0,0 +1,130 @@
+    private eventEmitter: EventEmitter2,
+  ) {
+    super();
+    this.logger = new Logger(this.constructor.name);
+  }
+
+  public async onApplicationBootstrap() {
+    this.logger.debug('Starting publishing service');
+  }
+
+  public onModuleDestroy() {
+    try {
+      this.logger.debug('Shutting down publishing service');
+    } catch (e) {
+      // 🐂 //
+    }
+  }
+
+  async process(job: Job<ITxMonitorJob, any, string>): Promise<any> {
+    this.logger.log(`Monitoring job ${job.id} of type ${job.name}`);
+    try {
+      const numberBlocksToParse = 100n;
+      const txCapacityEpoch = job.data.epoch;
+      const previousKnownBlockNumber = (await this.blockchainService.getBlock(job.data.lastFinalizedBlockHash)).block.header.number.toBigInt();
+      const currentFinalizedBlockNumber = await this.blockchainService.getLatestFinalizedBlockNumber();
+      const blockList: bigint[] = [];
+      for (let i = previousKnownBlockNumber; i <= currentFinalizedBlockNumber && i < previousKnownBlockNumber + numberBlocksToParse; i += 1n) {
+        blockList.push(i);
+      }
+      const txBlockHash = await this.crawlBlockList(job.data.txHash, txCapacityEpoch, blockList);
+
+      if (txBlockHash) {
+        this.logger.verbose(`Successfully completed job ${job.id}`);
+        return { success: true };
+      }
+      return { success: false };
+    } catch (e) {
+      this.logger.error(`Job ${job.id} failed (attempts=${job.attemptsMade}) with error: ${e}`);
+      throw e;
+    } finally {
+      // do some stuff
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  @OnWorkerEvent('completed')
+  onCompleted() {
+    // do some stuff
+  }
+
+  private async crawlBlockList(txHash: Hash, epoch: string, blockList: bigint[]): Promise<BlockHash | undefined> {
+    // Use Promise.all to await all filter operations before returning undefined
+    await Promise.all(
+      blockList.map(async (blockNumber) => {
+        const blockHash = await this.blockchainService.getBlockHash(blockNumber);
+        const block = await this.blockchainService.getBlock(blockHash);
+        const txInfo = block.block.extrinsics.filter((extrinsic) => extrinsic.hash.toString() === txHash.toString());
+        this.logger.debug(`Extrinsics: ${block.block.extrinsics[0]}`);
+
+        if (txInfo.length > 0) {
+          this.logger.debug(`Found tx ${txHash} in block ${blockNumber}`);
+          const at = await this.blockchainService.api.at(blockHash.toHex());
+          const events = await at.query.system.events();
+          events.subscribe((records) => {
+            records.forEach(async (record) => {
+              const { event } = record;
+              const eventName = event.section;
+              const { method } = event;
+              const { data } = event;
+              this.logger.debug(`Received event: ${eventName} ${method} ${data}`);
+              if (eventName.search('capacity') !== -1 && method.search('Withdrawn') !== -1) {
```

Comment:
should this be a utility function that detects capacity events?

Reply:
I think line 99-103 can be generalized in a way that you can pass a condition that checks the event and a function to run if that condition satisfied but that's a refactor that's not necessary at this point

Reply:
I think we can make it generic and move to blockchain service as a helper function